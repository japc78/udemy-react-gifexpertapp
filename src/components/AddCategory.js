import React, { useState } from 'react';
import PropTypes from 'prop-types';

export const AddCategory = ( { setCategories } ) => {

	const [inputValue, setInputValue] = useState('');

	const handleInputChange = (e) => {
		setInputValue(e.target.value);
		// console.log(e.target);
		// Comprobamos a la hora de hacer test para por el manejador
		// console.log('handleInputChange llamado');
	}

	const handleSubmit = (e) => {
		e.preventDefault();
		// Comprobamos a la hora de hacer test que simula el submit
		// console.log('Pass: handleSubmit', inputValue);

		if (inputValue.trim().length > 2) {
			setCategories(c => [inputValue, ...c]);
			setInputValue('');
			// console.log('Submit hecho!');
		}

	}

	return (
		<form onSubmit= {handleSubmit}>
			{/* Se a√±ade para comprobar que el valor del input cambio en el evento onChange */}
			<p>{ inputValue }</p>
			<h2>Add Category</h2>
			<input
				type="text"
				value = {inputValue}
				onChange = {handleInputChange}
			/>
		</form>
	)
}

AddCategory.propTypes = {
	setCategories : PropTypes.func.isRequired
}
