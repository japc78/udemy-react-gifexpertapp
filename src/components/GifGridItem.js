import React from 'react'

// Con props ya pasadas de manera independiente.
// export const GifGridItem = ( props ) => {
// 	console.log(props);

// De una manera mas amigable para la lectura del codigo
export const GifGridItem = ( {id, title, url} ) => {
	console.log({id, title, url});

	return (
		<div>
			<img src={url} alt={title}/>
			<p>{title}</p>
		</div>
	)
}
