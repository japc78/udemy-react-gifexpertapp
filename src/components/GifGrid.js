import React from 'react'
import { useFetchGifs } from '../hooks/useFetchGifs'
import { GifGridItem } from './GifGridItem';
import PropTypes from 'prop-types';

export const GifGrid = ( {category }) => {

	// Con desestructuracion
	const { data:images, loading} = useFetchGifs( category );

	return (
		<>
			<h3 className="animate__animated animate__fadeIn">{ category }</h3>

			{/* Operador ternario, si loading true se muestra cargando*/}
			{/* { loading ? 'Cargando...' : 'Datos cargados'} */}

			{ loading && <p className="animate__animated animate__flash">Cargando...</p>}


			<div className="card-grid">
				{
					// images.map(img => (
					// 	<li key = {img.id}> {img.title} </li>
					// ))

					// Con destructuracion
					images.map( img => (
						<GifGridItem
							key = {img.id}
							// img = {img}

							// Se envian las props diretamente con desestructuracion de manera independiente
							{...img}
						/>
					))}
			</div>

		</>
	)
}

GifGrid.propTypes = {
	category : PropTypes.string.isRequired
}