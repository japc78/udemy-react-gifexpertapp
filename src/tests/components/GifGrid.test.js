import React from 'react';
import '@testing-library/jest-dom'

import { shallow } from 'enzyme'
import { GifGrid } from '../../components/GifGrid';
import { useFetchGifs } from '../../hooks/useFetchGifs';
jest.mock('../../hooks/useFetchGifs')

describe('Purebas en componente <GifGrid />', () => {

	const category = 'Rick and Morty';

	test('Debe de mostrarse correctamente', () => {

		useFetchGifs.mockReturnValue({
			data: [],
			loading: true
		})

		const wrapper = shallow(<GifGrid category = {category} />);
		expect( wrapper ).toMatchSnapshot();
	});

	test('Debe de mostrar items cuando se cargan imagenes useFechGifs', () => {

		const gifs = [
			{
				id: 'ABC',
				url: 'https://localhost/testing/url.jpg',
				title: 'Elemento de prueba'
			},
			{
				id: 'DEF',
				url: 'https://localhost/testing/url2.jpg',
				title: 'Elemento de prueba 2'
			}
		]

		useFetchGifs.mockReturnValue({
			data: gifs,
			loading: false
		})

		const wrapper = shallow(<GifGrid category = {category} />);
		expect( wrapper ).toMatchSnapshot();

		// Se comprueba que el loading no exista
		expect( wrapper.find('p').exists()).toBe(false);

		// Se comprueba que se pinta dos componentes del GifGridItem que coincide con el array gifs de prueba
		expect( wrapper.find('GifGridItem').length).toBe(gifs.length);

	})

})
