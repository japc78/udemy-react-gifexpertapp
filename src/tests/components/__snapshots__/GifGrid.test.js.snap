// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Purebas en componente <GifGrid /> Debe de mostrar items cuando se cargan imagenes useFechGifs 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Rick and Morty
  </h3>
  <div
    className="card-grid"
  >
    <GifGridItem
      id="ABC"
      key="ABC"
      title="Elemento de prueba"
      url="https://localhost/testing/url.jpg"
    />
    <GifGridItem
      id="DEF"
      key="DEF"
      title="Elemento de prueba 2"
      url="https://localhost/testing/url2.jpg"
    />
  </div>
</Fragment>
`;

exports[`Purebas en componente <GifGrid /> Debe de mostrarse correctamente 1`] = `
<Fragment>
  <h3
    className="animate__animated animate__fadeIn"
  >
    Rick and Morty
  </h3>
  <p
    className="animate__animated animate__flash"
  >
    Cargando...
  </p>
  <div
    className="card-grid"
  />
</Fragment>
`;
